### ISSUE TITLE
Error with Dynamic Web Content

### ISSUE THREAD

**Issue by insetic** [mautic/mautic#2105]
What type of report is this:

| Q | A |
| --- | --- |
| Bug report? | Yes |
## Description:

Dynamic Content is not working properly in my fresh installation of Mautic. It sends the dynamic content at the first time but then generates an error that produce that the pushed dynamic content is no more updated.

Also, when I try to delete the dynamic content or the contact who see the content, they are not deleted and I get some errors in the log file.
## If a bug:

| Q | A |
| --- | --- |
| Mautic version | 2.01 |
| PHP version | 5.6.23 |
### Steps to reproduce:
1. Create a campaign that push a dynamic content to a contacts segment
2. When a contact show the dynamic content in the site, try to change the dynamic content to push: it doesn't work
3. Also try to delete the contact or the dynamic content that the contact has seen: they aren't deleted and in the log are registered this errors.
### Log errors:

[2016-07-25 09:14:30] mautic.CRITICAL: Uncaught PHP Exception InvalidArgumentException: "mautic.dwc.model.dwc is not a registered container key." at /var/www/vhosts/mysite.com/httpdocs/mautic/app/bundles/CoreBundle/Controller/CommonController.php line 120 {"exception":"[object](InvalidArgumentException%28code: 0%29: mautic.dwc.model.dwc is not a registered container key. at /var/www/vhosts/mysite.com/httpdocs/mautic/app/bundles/CoreBundle/Controller/CommonController.php:120)"} []
[2016-07-25 09:14:34] mautic.CRITICAL: Uncaught PHP Exception Doctrine\DBAL\Exception\ForeignKeyConstraintViolationException: "An exception occurred while executing 'DELETE FROM dynamic_content WHERE id = ?' with params [1]:  SQLSTATE[23000]: Integrity constraint violation: 1451 Cannot delete or update a parent row: a foreign key constraint fails (`mauticv2`.`dynamic_content_lead_data`, CONSTRAINT `FK_515B221BD9D0CD7` FOREIGN KEY (`dynamic_content_id`) REFERENCES `dynamic_content` (`id`))" at /var/www/vhosts/mysite.com/httpdocs/mautic/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractMySQLDriver.php line 60 {"exception":"[object](Doctrine\DBAL\Exception\ForeignKeyConstraintViolationException%28code: 0%29: An exception occurred while executing 'DELETE FROM dynamic_content WHERE id = ?' with params [1]:nnSQLSTATE[23000]: Integrity constraint violation: 1451 Cannot delete or update a parent row: a foreign key constraint fails %28`mauticv2`.`dynamic_content_lead_data`, CONSTRAINT `FK_515B221BD9D0CD7` FOREIGN KEY %28`dynamic_content_id`) REFERENCES `dynamic_content` (`id`)) at /var/www/vhosts/mysite.com/httpdocs/mautic/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractMySQLDriver.php:60, Doctrine\DBAL\Driver\PDOException(code: 23000): SQLSTATE[23000]: Integrity constraint violation: 1451 Cannot delete or update a parent row: a foreign key constraint fails (`mauticv2`.`dynamic_content_lead_data`, CONSTRAINT `FK_515B221BD9D0CD7` FOREIGN KEY (`dynamic_content_id`) REFERENCES `dynamic_content` (`id`)) at /var/www/vhosts/mysite.com/httpdocs/mautic/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php:93, PDOException(code: 23000): SQLSTATE[23000]: Integrity constraint violation: 1451 Cannot delete or update a parent row: a foreign key constraint fails (`mauticv2`.`dynamic_content_lead_data`, CONSTRAINT `FK_515B221BD9D0CD7` FOREIGN KEY (`dynamic_content_id`) REFERENCES `dynamic_content` (`id`)) at /var/www/vhosts/mysite.com/httpdocs/mautic/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php:91)"} []

[2016-07-25 09:13:41] mautic.CRITICAL: Uncaught PHP Exception Doctrine\DBAL\Exception\ForeignKeyConstraintViolationException: "An exception occurred while executing 'DELETE FROM leads WHERE id = ?' with params [2]:  SQLSTATE[23000]: Integrity constraint violation: 1451 Cannot delete or update a parent row: a foreign key constraint fails (`mauticv2`.`dynamic_content_lead_data`, CONSTRAINT `FK_515B221B55458D` FOREIGN KEY (`lead_id`) REFERENCES `leads` (`id`))" at /var/www/vhosts/mysite.com/httpdocs/mautic/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractMySQLDriver.php line 60 {"exception":"[object](Doctrine\DBAL\Exception\ForeignKeyConstraintViolationException%28code: 0%29: An exception occurred while executing 'DELETE FROM leads WHERE id = ?' with params [2]:nnSQLSTATE[23000]: Integrity constraint violation: 1451 Cannot delete or update a parent row: a foreign key constraint fails %28`mauticv2`.`dynamic_content_lead_data`, CONSTRAINT `FK_515B221B55458D` FOREIGN KEY %28`lead_id`) REFERENCES `leads` (`id`)) at /var/www/vhosts/mysite.com/httpdocs/mautic/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractMySQLDriver.php:60, Doctrine\DBAL\Driver\PDOException(code: 23000): SQLSTATE[23000]: Integrity constraint violation: 1451 Cannot delete or update a parent row: a foreign key constraint fails (`mauticv2`.`dynamic_content_lead_data`, CONSTRAINT `FK_515B221B55458D` FOREIGN KEY (`lead_id`) REFERENCES `leads` (`id`)) at /var/www/vhosts/mysite.com/httpdocs/mautic/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php:93, PDOException(code: 23000): SQLSTATE[23000]: Integrity constraint violation: 1451 Cannot delete or update a parent row: a foreign key constraint fails (`mauticv2`.`dynamic_content_lead_data`, CONSTRAINT `FK_515B221B55458D` FOREIGN KEY (`lead_id`) REFERENCES `leads` (`id`)) at /var/www/vhosts/mysite.com/httpdocs/mautic/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOStatement.php:91)"} []

**Comment by insetic** [comment]
Hi @dongilbert @dbhurley 
Do you have any idea about when will be resolved this issue?
We're eager to use this Mautic feature.
Thanks !!

**Comment by insetic** [comment]
Hi again @dongilbert , in slack you said this 20 days ago. What could I do to get this feature working...? Thanks !!!

>  dongilbert
> We are working on a fix for that. There are certain scenarios where Dynamic Content will fire and try to load the data prior to the lead cookie being populated, which would explain why it loads once and then not again.
> For your second issue, we need to update the db column definition on a couple items so that they automatically delete themselves or set to NULL when the parent is deleted, so as to prevent the foreign constraint error
> Posted in #supportAug 2nd at 5:10 PM

**Comment by dongilbert** [comment]
@insetic - The fix for that should be in 2.1.1. I'll have to double check.

**Comment by insetic** [comment]
hi @dongilbert 
I've tested 2.1.1 (in a new, fresh Mautic installation to avoid errors) and it doesn't work. 
I get the same wrong behavior with Dynamic Web Content, and also the same error when I try to delete a lead who has view a DWC.

**Comment by insetic** [comment]
Hi again @dongilbert 
We're waiting for a solution over a month and now it's urgent for us to use this feature.
As you said, this problem occurs in "certain scenarios", would you describe an scenario that we can reproduce to get the DWC working properly?

**Comment by Mazzim** [comment]
@insetic yes, Dynamic Web Content works for me on Mautic 2.1.1. Unknown contacts see one content, known ones see another, both at first DWC request and subsequent DWC pull requests.
Which version are you on at the moment?

As for the error, I had the same 'FOREIGN KEY CONSTRAINT' one in my logs once, when trying to delete a contact that has DWC served to. `dynamic_content_lead_data` table is referenced by `leads` table, and should have 'ON DELETE CASCADE' set after recent updates that include DB migrations. When you delete a lead, the corresponding row from  `dynamic_content_lead_data` should be deleted automatically. It doesn't, so delete contact (lead) fails also, and the error is thrown in in to logs. This suggests that you do not have that [particular DB migration](https://github.com/mautic/mautic/blob/e0144b44a2380179acd93e122d83b80ee4f14534/app/migrations/Version20160617000318.php) or for some reason it failed.

Your first check would be DB migrations (`php app/console doctrine:migration:status` and if you see New migrations shown, `php app/console doctrine:migration:migrate`)
I had 0 new migrations but still getting that error. So I had to manually add 'ON DELETE CASCADE' (see two last lines of the DB migration I referenced). 

If you have command line access to your server, here is a step-by-step:

- Login to your mysql or use phpMyAdmin SQL tab
- Find out the FOREIGN_KEY for dynamic_content_lead_data

`USE mautic;`
(make sure you change 'mautic' to your actual DB name)
`SHOW CREATE TABLE dynamic_content_lead_data;`
mark down the foreign key against dynamic_content_lead_data, 
- Drop that foreign key

`ALTER TABLE dynamic_content_lead_data DROP FOREIGN KEY 'FK_515B221B55458D';`

- Add foreign key back in with  'ON DELETE CASCADE'

`ALTER TABLE dynamic_content_lead_data
   ADD CONSTRAINT 'FK_515B221B55458D'
   FOREIGN KEY ('lead_id')
   REFERENCES 'leads' ('id' )
   ON DELETE CASCADE;`

- Check if you do not have 'ON DELETE CASCADE' for dynamic_content table also, and if you don't, repeat the above steps, replacing `leads` with `dynamic_content` and `lead_id` with `dynamic_content_id`

- You should be able to delete both contacts and dynamic content without errors

Hope this helps!

**Comment by Mazzim** [comment]
@insetic on your main issue, not sure if [this](https://github.com/mautic/mautic/issues/2004) is related, but are your cron jobs staggered apart?

**Comment by insetic** [comment]
Thanks @Mazzim, your help is really appreciated !! It resolves the second part of my problem.

The first one is about this:
Have you tested what happens if you change, in the campaign, the dynamic content that see a contact (known or not)? 
In my case, if I change the dynamic content, the contact still see the first content and the new content never is pushed...

@dongilbert said this in relation to the problem:
> There are certain scenarios where Dynamic Content will fire and try to load the data prior to the lead cookie being populated, which would explain why it loads once and then not again.

**Comment by Mazzim** [comment]
Glad it helped with the 2nd part :)

With the actual issue,
> Have you tested what happens if you change, in the campaign, the dynamic content that see a contact (known or not)? 
In my case, if I change the dynamic content, the contact still see the first content and the new content never is pushed...

I'm sorry, I am not sure I'm clear on the steps to reproduce it. You wrote:

> Create a campaign that push a dynamic content to a contacts segment
When a contact show the dynamic content in the site, try to change the dynamic content to push: it doesn't work

Do you mean that
- you have a campaign running where you have an action 'Request dynamic content' with, say 'slotA' and 'dynamicContentA' chosen
- when your visitors open a page with the DWC code for the first time, if it's unknown visitor, a default content from the code is shown. For a known contact content from 'dynamicCoontentA' is shown, right?

Now you propose to change the dynamic content. Do you mean actually going to dynamicCoontentA, opening it and changing the content itself? Or do you have a condition block that follows your 'Request dynamic content' block, with 'Push dynamic content' attached to a green/positive anchor that suppose to present different content to a contact if the condition is met? Can you please elaborate on it a bit? Or upload a screenshot or two?

**Comment by insetic** [comment]
Yes, just go to dynamicCoontentA, open it and change the content itself.
In my case, the contact still see the old dynamicCoontentA instead of the new one.

**Comment by insetic** [comment]
Thanks a lot @Mazzim, the second part of the problem was resolved following your instructions.
As a summary, anyone who has the same problem, the only thing that must do is, from his PHPMyAdmin, execute this:
```
ALTER TABLE dynamic_content_lead_data DROP FOREIGN KEY `FK_515B221B55458D`;
ALTER TABLE dynamic_content_lead_data ADD CONSTRAINT `FK_515B221B55458D` FOREIGN KEY (`lead_id`) REFERENCES `leads` (`id`) ON DELETE CASCADE;
ALTER TABLE dynamic_content_lead_data DROP FOREIGN KEY `FK_515B221BD9D0CD7`;
ALTER TABLE dynamic_content_lead_data ADD CONSTRAINT `FK_515B221BD9D0CD7` FOREIGN KEY (`dynamic_content_id`) REFERENCES `dynamic_content` (`id`) ON DELETE CASCADE;
```

**Comment by Mazzim** [comment]
@insetic I've tried changing the dynamic content after it has been shown to a known visitor for the first time. Then I tried opening my page with my DWC code. I tried it both with Request dynamic content block and with Push dynamic content, once I refresh a page with my DWC code, it shows the new dynamic content that I just edited. I see no issues, it is an expected behavior.

**Comment by insetic** [comment]
Thanks for your time and your interest @Mazzim 
As @dongilbert said, it happens in certain scenarios, so I'll wait for the fix. Unfortunately my scenario doesn't work as expected...

**Comment by dongilbert** [comment]
The scenario we discovered that would cause dynamic content not to function properly is when the Tracking Pixel was loaded after the dynamic content replacements had already fired, causing the contact to not be identified by the time the replacements were made. On subsequent requests, the contact would be identified because the cookie is present, and would be shown potentially different dynamic content.

**Comment by insetic** [comment]
thanks @dongilbert, that probably explains the malfunction that occurs in my platform.

**Comment by brandner** [comment]
Similar to this issue, unable to delete contacts that have related dynamic content...

`mautic.CRITICAL: Uncaught PHP Exception Doctrine\DBAL\Exception\ForeignKeyConstraintViolationException: "An exception occurred while executing 'DELETE FROM leads WHERE id = ?' with params [2]:  SQLSTATE[23000]: Integrity constraint violation: 1451 Cannot delete or update a parent row: a foreign key constraint fails (`c0mautic`.`dynamic_content_lead_data`, CONSTRAINT `FK_515B221B55458D` FOREIGN KEY (`lead_id`) REFERENCES `leads` (`id`))" at /var/www/clients/client0/web2/web/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractMySQLDriver.php line 60 `

**Comment by wittwerch** [comment]
@brandner can confirm that bug on 2.4.0. Contacts that received dynamic content can't be deleted anymore.

**Comment by brandner** [comment]
For me at least, this bug is fixed with release 2.6.0.
